beacon_node/rest_api/src/validator.rs

```diff
@@ -0,0 +1,149 @@
+use types::{BeaconBlock, BeaconState, Epoch, RelativeEpoch, Shard, Slot};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ValidatorDuty {
+    /// The validator's BLS public key, uniquely identifying them. _48-bytes, hex encoded with 0x prefix, case insensitive._
+    #[serde(rename = "validator_pubkey")]
+    pub validator_pubkey: String,
+    /// The slot at which the validator must attest.
+    #[serde(rename = "attestation_slot")]
+    pub attestation_slot: Option<Slot>,
+    /// The shard in which the validator must attest.
+    #[serde(rename = "attestation_shard")]
+    pub attestation_shard: Option<Shard>,
+    /// The slot in which a validator must propose a block, or `null` if block production is not required.
+    #[serde(rename = "block_proposal_slot")]
+    pub block_proposal_slot: Option<Slot>,
+}
+
+impl ValidatorDuty {
+    pub fn new() -> ValidatorDuty {
+        ValidatorDuty {
+            validator_pubkey: "".to_string(),
+            attestation_slot: None,
+            attestation_shard: None,
+            block_proposal_slot: None,
+        }
+    }
+}
+
+/// HTTP Handler to retrieve a the duties for a set of validators during a particular epoch
+pub fn get_validator_duties<T: BeaconChainTypes + 'static>(req: Request<Body>) -> ApiResult {
+    // Get beacon state
+    let beacon_chain = req
+        .extensions()
+        .get::<Arc<BeaconChain<T>>>()
+        .ok_or_else(|| ApiError::ServerError("Beacon chain extension missing".to_string()))?;
+    let _ = beacon_chain
+        .ensure_state_caches_are_built()
+        .map_err(|e| ApiError::ServerError(format!("Unable to build state caches: {:?}", e)))?;
+    let head_state = beacon_chain
+        .speculative_state()
+        .expect("This is legacy code and should be removed.");
+
+    // Parse and check query parameters
+    let query = UrlQuery::from_request(&req)?;
+
+    let current_epoch = head_state.current_epoch();
+    let epoch = match query.first_of(&["epoch"]) {
+        Ok((_, v)) => Epoch::new(v.parse::<u64>().map_err(|e| {
+            ApiError::InvalidQueryParams(format!("Invalid epoch parameter, must be a u64. {:?}", e))
+        })?),
+        Err(_) => {
+            // epoch not supplied, use the current epoch
+            current_epoch
+        }
+    };
+    let relative_epoch = RelativeEpoch::from_epoch(current_epoch, epoch)
+        .map_err(|e| ApiError::InvalidQueryParams(format!("Cannot get RelativeEpoch: {:?}", e)))?;
+    //TODO: Handle an array of validators, currently only takes one
+    let mut validators: Vec<PublicKey> = match query.all_of("validator_pubkeys") {
+        Ok(v) => v
+            .iter()
+            .map(|pk| parse_pubkey(pk))
+            .collect::<Result<Vec<_>, _>>()?,
+        Err(e) => {
+            return Err(e);
+        }
+    };
+    let mut duties: Vec<ValidatorDuty> = Vec::new();
+
+    // Get a list of all validators for this epoch
+    let validator_proposers: Result<Vec<usize>, _> = epoch
```

Comment:
I think lines 81..93 could become:

```rust
    // Get a list of all validators for this epoch
    let validator_proposers: Result<Vec<usize>, _> = epoch
        .slot_iter(T::EthSpec::slots_per_epoch())
        .map(|slot| head_state.get_beacon_proposer_index(slot, relative_epoch, &beacon_chain.spec))
        .collect()
        .map_err(|e| ApiError:: ... )?;
```

Reply:
Cool, thanks. I've reduced it to something, similar to your suggestion:
```
// Get a list of all validators for this epoch
    let validator_proposers: Vec<usize> = epoch
        .slot_iter(T::EthSpec::slots_per_epoch())
        .map(|slot| {
            head_state
                .get_beacon_proposer_index(slot, relative_epoch, &beacon_chain.spec)
                .map_err(|e| {
                    ApiError::ServerError(format!(
                        "Unable to get proposer index for validator: {:?}",
                        e
                    ))
                })
        })
        .collect::<Result<Vec<usize>, _>>()?;
```